File: haushaltsbuch/src/main/java/org/github/haschi/haushaltsbuch/api/BuchungConverter.java
package org.github.haschi.haushaltsbuch.api;

import cucumber.api.Transformer;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressWarnings("WeakerAccess")
public class BuchungConverter extends Transformer<Buchung>
{
    @Override
    public Buchung transform(final String s)
    {
        final Pattern pattern = Pattern.compile("^(.*) (-?(?:\\d{1,3}\\.)?\\d{1,3},\\d{2} EUR)$");
        final Matcher matcher = pattern.matcher(s);

        assert matcher.matches() : "Keine Übereinstimmung: " + s;

        if (matcher.matches())
        {
            return Buchung.builder()
                    .buchungstext(matcher.group(1))
                    .betrag(Währungsbetrag.parse(matcher.group(2)))
                    .build();
        }
        else
        {
            return _Buchung.leer();
        }
    }
}

File: testit/src/test/kotlin/com/github/haschi/domain/haushaltsbuch/testing/BuchungConverter.kt
package com.github.haschi.domain.haushaltsbuch.testing

import cucumber.api.Transformer
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Währungsbetrag
import com.github.haschi.dominium.haushaltsbuch.core.model.values.Buchung
import java.util.regex.Pattern

class BuchungConverter : Transformer<Buchung>()
{

    override fun transform(s: String): Buchung
    {
        val pattern = Pattern.compile(MoneyConverter.regex)
        val matcher = pattern.matcher(s)

        assert(matcher.matches()) { "Keine Übereinstimmung: " + s }

        return if (matcher.matches())
        {
            Buchung(
                    buchungstext = matcher.group(1),
                    betrag = Währungsbetrag.währungsbetrag(matcher.group(2)))
        } else
        {
            Buchung.leer
        }
    }
}

